lib/MySQL_Session.cpp

```diff
@@ -348,15 +550,571 @@ int MySQL_Session::handler() {
 
-__get_a_backend:
+handler_again:
+
+	switch (status) {
+		case FAST_FORWARD:
+			fprintf(stderr,"FAST_FORWARD\n");
+			// FIXME: to implement
+			break;
+		case CONNECTING_CLIENT:
+			//fprintf(stderr,"CONNECTING_CLIENT\n");
+			// FIXME: to implement
+			break;
+		case PINGING_SERVER:
+			
+/*
+			if (mybe->server_myds->revents) {
+				MySQL_Data_Stream *myds=mybe->server_myds;
+				MySQL_Connection *myconn=myds->myconn;
+				myconn->handler(myds->revents);
+				if (myconn->async_state_machine==ASYNC_PING_SUCCESSFUL) {
+					myds->DSS=STATE_READY;
+					myds->myconn->async_state_machine=ASYNC_IDLE;
+					if ((myds->myconn->reusable==true) && ((myds->myprot.prot_status & SERVER_STATUS_IN_TRANS)==0)) {
+
+						return_MySQL_Connection_To_Poll(myds);
+					}
+					status=NONE;
+				} else {
+					if (myconn->async_state_machine==ASYNC_PING_FAILED) {
+						// FIXME: treat gracefully
+						assert(0);
+					}
+				}
+			}
+*/
+			assert(mybe->server_myds->myconn);
+			{	
+				MySQL_Data_Stream *myds=mybe->server_myds;
+				MySQL_Connection *myconn=myds->myconn;
+				int rc=myconn->async_ping(myds->revents);
+//				if (myds->mypolls==NULL) {
+//					thread->mypolls.add(POLLIN|POLLOUT, myds->fd, myds, thread->curtime);
+//				}
+				if (rc==0) {
+					myconn->async_state_machine=ASYNC_IDLE;
+					//if ((myconn->reusable==true) && ((myds->myprot.prot_status & SERVER_STATUS_IN_TRANS)==0)) {
+					if ((myconn->reusable==true) && myds->myconn->IsActiveTransaction()==false) {
+						myds->return_MySQL_Connection_To_Pool();
+					}
+					delete mybe->server_myds;
+					mybe->server_myds=NULL;
+					set_status(NONE);
+					return -1;
+				} else {
+					if (rc==-1) {
+						proxy_error("Detected a broken connection during ping\n");
+						myds->destroy_MySQL_Connection_From_Pool();
+						myds->fd=0;
+						delete mybe->server_myds;
+						mybe->server_myds=NULL;
+						//thread->mypolls.remove_index_fast(myds->poll_fds_idx);
+						return -1;
+					} else {
+						// rc==1 , nothing to do for now
+// tring to fix bug
+						if (myds->mypolls==NULL) {
+							thread->mypolls.add(POLLIN|POLLOUT, myds->fd, myds, thread->curtime);
+						}
+// tring to fix bug
+					}
+				}
+			}
+			break;
+
+		case PROCESSING_QUERY:
+			//fprintf(stderr,"PROCESSING_QUERY\n");
+			if (pause_until > thread->curtime) {
+				return 0;
+			}
+			if (mysql_thread___connect_timeout_server_max) {
+				if (mybe->server_myds->max_connect_time==0)
+					mybe->server_myds->max_connect_time=thread->curtime+mysql_thread___connect_timeout_server_max*1000;
+			} else {
+				mybe->server_myds->max_connect_time=0;
+			}
+			if (
+				(mybe->server_myds->myconn && mybe->server_myds->wait_until && thread->curtime >= mybe->server_myds->wait_until)
+				// query timed out
+				||
+				(killed==true) // session was killed by admin
+			) {
+				MySQL_Data_Stream *myds=mybe->server_myds;
+				if (myds->myconn && myds->myconn->mysql) {
+					if (myds->killed_at==0) {
+						myds->wait_until=0;
```

Comment:
Did I understand correctly - you're doing it in a separate thread because you wanted to use synchronous primitives for mysql_connect, mysql_query, etc? We should document this in a comment here.